File: engines/go/internal/render/render.go
package render

import (
	"fmt"
	"math"
	"os"
	"path/filepath"
	"regexp"
	"strings"

	"statusline/internal/cache"
	"statusline/internal/config"
	"statusline/internal/format"
	"statusline/internal/gitstate"
	"statusline/internal/session"
)

const (
	dim     = "\033[2m"
	rst     = "\033[0m"
	cyan    = "\033[36m"
	green   = "\033[32m"
	yellow  = "\033[33m"
	red     = "\033[31m"
	magenta = "\033[35m"
)

var ansiRe = regexp.MustCompile(`\033\[[0-9;]*m`)

// Render produces the 2-line statusline output from session data and config.
func Render(sess session.Session, cfg config.Config) string {
	sep := " " + dim + "\u2502" + rst + " "

	// Session ID from transcript path
	sessionID := ""
	if sess.TranscriptPath != "" {
		base := filepath.Base(sess.TranscriptPath)
		sessionID = strings.TrimSuffix(base, ".jsonl")
	}

	// --- Model ---
	model := ""
	if cfg.ShowModel {
		model = sess.Model.DisplayName
		if model == "" {
			model = "?"
		}
		model = strings.TrimPrefix(model, "Claude ")
	}

	// --- Context bar ---
	pct := 0
	bar := ""
	warn := ""
	clr := green
	if cfg.ShowContext {
		pct = int(sess.ContextWindow.UsedPercentage)
		filled := pct / 10
		if filled > 10 {
			filled = 10
		}
		empty := 10 - filled
		bar = strings.Repeat("\u2593", filled) + strings.Repeat("\u2591", empty)
		if pct >= 90 {
			clr = red
			warn = " \u26a0"
		} else if pct >= 70 {
			clr = yellow
			warn = " \u26a0"
		}
	}

	// --- Cost ---
	costFmt := ""
	if cfg.ShowCost {
		costFmt = format.FmtCost(sess.Cost.TotalCostUSD)
	}

	// --- Duration ---
	durFmt := ""
	if cfg.ShowDuration {
		durFmt = format.FmtDuration(int(sess.Cost.TotalDurationMs))
	}

	// --- Git ---
	var gitDisplay, dirty, gitExtra string
	if cfg.ShowGit {
		cwd, err := os.Getwd()
		if err == nil {
			gs := gitstate.Get(cwd)
			if gs != nil && gs.Branch != "" {
				sb := format.Truncate(format.ShortenBranch(gs.Branch), 20)
				if gs.InWorktree {
					sw := format.Truncate(format.ShortenBranch(gs.WorktreeName), 20)
					if sw == sb {
						gitDisplay = "\u2295 " + sb
					} else {
						gitDisplay = "\u2295" + sw + " " + sb
					}
				} else {
					gitDisplay = sb
				}
				if gs.Dirty {
					dirty = "\u25cf"
				}
				var parts []string
				if gs.Ahead > 0 {
					parts = append(parts, fmt.Sprintf("\u2191%d", gs.Ahead))
				}
				if gs.Behind > 0 {
					parts = append(parts, fmt.Sprintf("\u2193%d", gs.Behind))
				}
				if gs.Stash > 0 {
					parts = append(parts, fmt.Sprintf("stash:%d", gs.Stash))
				}
				gitExtra = strings.Join(parts, " ")
			}
		}
	}

	// --- Lines added/removed ---
	linesFmt := ""
	if cfg.ShowDiff {
		added := int(sess.Cost.TotalLinesAdded)
		removed := int(sess.Cost.TotalLinesRemoved)
		if added > 0 || removed > 0 {
			linesFmt = fmt.Sprintf("%s+%d%s %s-%d%s", green, added, rst, red, removed, rst)
		}
	}

	// --- Token data ---
	inTok := int(sess.ContextWindow.TotalInputTokens)
	outTok := int(sess.ContextWindow.TotalOutputTokens)
	inFmt := format.FmtK(inTok)
	outFmt := format.FmtK(outTok)

	// --- Per-model stats ---
	var modelStats *cache.ModelStats
	modelMix := ""
	if sessionID != "" {
		modelStats = cache.ReadModels(sessionID)
	}

	if modelStats != nil {
		maxOut := modelStats.OpusOut
		if modelStats.SonnetOut > maxOut {
			maxOut = modelStats.SonnetOut
		}
		if modelStats.HaikuOut > maxOut {
			maxOut = modelStats.HaikuOut
		}
		if cfg.ShowModelBars && maxOut > 0 {
			oBar := format.BarChar(modelStats.OpusOut, maxOut)
			sBar := format.BarChar(modelStats.SonnetOut, maxOut)
			hBar := format.BarChar(modelStats.HaikuOut, maxOut)
			oC := dim + "·"
			if oBar != "" {
				oC = magenta + oBar
			}
			sC := dim + "·"
			if sBar != "" {
				sC = cyan + sBar
			}
			hC := dim + "·"
			if hBar != "" {
				hC = green + hBar
			}
			modelMix = oC + sC + hC + rst
		}
	}

	// --- Speed ---
	speedFmt := ""
	if cfg.ShowSpeed {
		apiMs := int(sess.Cost.TotalAPIDurationMs)
		if apiMs > 0 && outTok > 0 {
			speed := float64(outTok) * 1000 / float64(apiMs)
			speedInt := int(math.RoundToEven(speed))
			speedClr := red
			if speedInt > 30 {
				speedClr = green
			} else if speedInt >= 15 {
				speedClr = yellow
			}
			speedFmt = fmt.Sprintf("%s%d tok/s%s", speedClr, speedInt, rst)
		}
	}

	// --- Cumulative stats ---
	cumProj := ""
	cumAll := ""
	if cfg.ShowCumulative {
		projStats, allStats := cache.ReadCumulative(sess.Workspace.ProjectDir)
		if projStats != nil {
			cumProj = fmt.Sprintf("\u2302 %s/%s/%s",
				format.FmtCost(projStats.D1),
				format.FmtCost(projStats.D7),
				format.FmtCost(projStats.D30))
		}
		if allStats != nil {
			cumAll = fmt.Sprintf("\u03a3 %s/%s/%s",
				format.FmtCost(allStats.D1),
				format.FmtCost(allStats.D7),
				format.FmtCost(allStats.D30))
		}
	}

	// ======== ASSEMBLE LINE 1 ========
	var l1Parts []string

	if model != "" {
		part := cyan + model + rst
		if modelMix != "" {
			part += " " + modelMix
		}
		l1Parts = append(l1Parts, part)
	} else if modelMix != "" {
		l1Parts = append(l1Parts, modelMix)
	}

	if bar != "" {
		l1Parts = append(l1Parts, fmt.Sprintf("%s%s %d%%%s%s", clr, bar, pct, warn, rst))
	}
	if costFmt != "" {
		l1Parts = append(l1Parts, costFmt)
	}
	if durFmt != "" {
		l1Parts = append(l1Parts, durFmt)
	}
	if gitDisplay != "" {
		gitPart := magenta + gitDisplay + rst
		if dirty != "" {
			gitPart += " " + yellow + dirty + rst
		}
		if gitExtra != "" {
			gitPart += " " + cyan + gitExtra + rst
		}
		l1Parts = append(l1Parts, gitPart)
	}
	if linesFmt != "" {
		l1Parts = append(l1Parts, linesFmt)
	}

	l1 := strings.Join(l1Parts, sep)

	// ======== ASSEMBLE LINE 2 ========
	l2 := ""
	if cfg.Line2 {
		var l2Parts []string

		if cfg.ShowTokens {
			var tokParts []string
			if modelStats != nil && (modelStats.OpusOut > 0 || modelStats.OpusIn > 0) {
				tokParts = append(tokParts, fmt.Sprintf("%sO%s:%s/%s",
					magenta, rst, format.FmtK(modelStats.OpusIn), format.FmtK(modelStats.OpusOut)))
			}
			if modelStats != nil && (modelStats.SonnetOut > 0 || modelStats.SonnetIn > 0) {
				tokParts = append(tokParts, fmt.Sprintf("%sS%s:%s/%s",
					cyan, rst, format.FmtK(modelStats.SonnetIn), format.FmtK(modelStats.SonnetOut)))
			}
			if modelStats != nil && (modelStats.HaikuOut > 0 || modelStats.HaikuIn > 0) {
				tokParts = append(tokParts, fmt.Sprintf("%sH%s:%s/%s",
					green, rst, format.FmtK(modelStats.HaikuIn), format.FmtK(modelStats.HaikuOut)))
			}
			if len(tokParts) > 0 {
				l2Parts = append(l2Parts, strings.Join(tokParts, " "))
			} else {
				l2Parts = append(l2Parts, fmt.Sprintf("%sin:%s%s %sout:%s%s",
					dim, rst, inFmt, dim, rst, outFmt))
			}
		}

		if speedFmt != "" {
			l2Parts = append(l2Parts, speedFmt)
		}
		if cumProj != "" {
			l2Parts = append(l2Parts, cumProj)
		}
		if cumAll != "" {
			l2Parts = append(l2Parts, cumAll)
		}

		l2 = strings.Join(l2Parts, sep)
	}

	// --- NO_COLOR ---
	if cfg.NoColor {
		l1 = ansiRe.ReplaceAllString(l1, "")
		l2 = ansiRe.ReplaceAllString(l2, "")
	}

	if l2 != "" {
		return l1 + "\n" + l2 + "\n"
	}
	return l1 + "\n\n"
}

File: engines/rust/src/render.rs
use crate::cache;
use crate::config::Config;
use crate::format;
use crate::session::Session;
use std::path::Path;

const DIM: &str = "\x1b[2m";
const RST: &str = "\x1b[0m";
const CYAN: &str = "\x1b[36m";
const GREEN: &str = "\x1b[32m";
const YELLOW: &str = "\x1b[33m";
const RED: &str = "\x1b[31m";
const MAGENTA: &str = "\x1b[35m";

pub fn render(sess: &Session, cfg: &Config) -> String {
    let sep = format!(" {}\u{2502}{} ", DIM, RST);

    // Session ID from transcript path
    let session_id = if !sess.transcript_path.is_empty() {
        let base = Path::new(&sess.transcript_path)
            .file_name()
            .map(|f| f.to_string_lossy().to_string())
            .unwrap_or_default();
        base.strip_suffix(".jsonl").unwrap_or(&base).to_string()
    } else {
        String::new()
    };

    // --- Model ---
    let model = if cfg.show_model {
        let m = if sess.model.display_name.is_empty() {
            "?".to_string()
        } else {
            sess.model.display_name.clone()
        };
        m.strip_prefix("Claude ").unwrap_or(&m).to_string()
    } else {
        String::new()
    };

    // --- Context bar ---
    let mut pct = 0i32;
    let mut bar = String::new();
    let mut warn = "";
    let mut clr = GREEN;
    if cfg.show_context {
        pct = sess.context_window.used_percentage as i32;
        let filled = (pct / 10).clamp(0, 10) as usize;
        let empty = 10 - filled;
        bar = "\u{2593}".repeat(filled) + &"\u{2591}".repeat(empty);
        if pct >= 90 {
            clr = RED;
            warn = " \u{26a0}";
        } else if pct >= 70 {
            clr = YELLOW;
            warn = " \u{26a0}";
        }
    }

    // --- Cost ---
    let cost_fmt = if cfg.show_cost {
        format::fmt_cost(sess.cost.total_cost_usd)
    } else {
        String::new()
    };

    // --- Duration ---
    let dur_fmt = if cfg.show_duration {
        format::fmt_duration(sess.cost.total_duration_ms as i64)
    } else {
        String::new()
    };

    // --- Git ---
    let mut git_display = String::new();
    let mut dirty = "";
    let mut git_extra = String::new();
    if cfg.show_git {
        if let Ok(cwd) = std::env::current_dir() {
            if let Some(gs) = crate::git::get(&cwd.to_string_lossy()) {
                if !gs.branch.is_empty() {
                    let sb = format::truncate(&format::shorten_branch(&gs.branch), 20);
                    if gs.in_worktree {
                        let sw = format::truncate(&format::shorten_branch(&gs.worktree_name), 20);
                        if sw == sb {
                            git_display = format!("\u{2295} {}", sb);
                        } else {
                            git_display = format!("\u{2295}{} {}", sw, sb);
                        }
                    } else {
                        git_display = sb;
                    }
                    if gs.dirty {
                        dirty = "\u{25cf}";
                    }
                    let mut parts = Vec::new();
                    if gs.ahead > 0 {
                        parts.push(format!("\u{2191}{}", gs.ahead));
                    }
                    if gs.behind > 0 {
                        parts.push(format!("\u{2193}{}", gs.behind));
                    }
                    if gs.stash > 0 {
                        parts.push(format!("stash:{}", gs.stash));
                    }
                    git_extra = parts.join(" ");
                }
            }
        }
    }

    // --- Lines added/removed ---
    let lines_fmt = if cfg.show_diff {
        let added = sess.cost.total_lines_added as i64;
        let removed = sess.cost.total_lines_removed as i64;
        if added > 0 || removed > 0 {
            format!("{}+{}{} {}-{}{}", GREEN, added, RST, RED, removed, RST)
        } else {
            String::new()
        }
    } else {
        String::new()
    };

    // --- Token data ---
    let in_tok = sess.context_window.total_input_tokens as i64;
    let out_tok = sess.context_window.total_output_tokens as i64;
    let in_fmt = format::fmt_k(in_tok);
    let out_fmt = format::fmt_k(out_tok);

    // --- Per-model stats ---
    let model_stats = if !session_id.is_empty() {
        cache::read_models(&session_id)
    } else {
        None
    };

    let model_mix = if let Some(ref ms) = model_stats {
        let max_out = ms.opus_out.max(ms.sonnet_out).max(ms.haiku_out);
        if cfg.show_model_bars && max_out > 0 {
            let o_bar = format::bar_char(ms.opus_out, max_out);
            let s_bar = format::bar_char(ms.sonnet_out, max_out);
            let h_bar = format::bar_char(ms.haiku_out, max_out);
            let o_c = if o_bar.is_empty() {
                format!("{}\u{00b7}", DIM)
            } else {
                format!("{}{}", MAGENTA, o_bar)
            };
            let s_c = if s_bar.is_empty() {
                format!("{}\u{00b7}", DIM)
            } else {
                format!("{}{}", CYAN, s_bar)
            };
            let h_c = if h_bar.is_empty() {
                format!("{}\u{00b7}", DIM)
            } else {
                format!("{}{}", GREEN, h_bar)
            };
            format!("{}{}{}{}", o_c, s_c, h_c, RST)
        } else {
            String::new()
        }
    } else {
        String::new()
    };

    // --- Speed ---
    let speed_fmt = if cfg.show_speed {
        let api_ms = sess.cost.total_api_duration_ms as i64;
        if api_ms > 0 && out_tok > 0 {
            let speed = out_tok as f64 * 1000.0 / api_ms as f64;
            let speed_int = format::round_to_even(speed);
            let speed_clr = if speed_int > 30 {
                GREEN
            } else if speed_int >= 15 {
                YELLOW
            } else {
                RED
            };
            format!("{}{} tok/s{}", speed_clr, speed_int, RST)
        } else {
            String::new()
        }
    } else {
        String::new()
    };

    // --- Cumulative stats ---
    let mut cum_proj = String::new();
    let mut cum_all = String::new();
    if cfg.show_cumulative {
        let (proj_stats, all_stats) = cache::read_cumulative(&sess.workspace.project_dir);
        if let Some(ps) = proj_stats {
            cum_proj = format!(
                "\u{2302} {}/{}/{}",
                format::fmt_cost(ps.d1),
                format::fmt_cost(ps.d7),
                format::fmt_cost(ps.d30)
            );
        }
        if let Some(als) = all_stats {
            cum_all = format!(
                "\u{03a3} {}/{}/{}",
                format::fmt_cost(als.d1),
                format::fmt_cost(als.d7),
                format::fmt_cost(als.d30)
            );
        }
    }

    // ======== ASSEMBLE LINE 1 ========
    let mut l1_parts: Vec<String> = Vec::new();

    if !model.is_empty() {
        let mut part = format!("{}{}{}", CYAN, model, RST);
        if !model_mix.is_empty() {
            part = format!("{} {}", part, model_mix);
        }
        l1_parts.push(part);
    } else if !model_mix.is_empty() {
        l1_parts.push(model_mix.clone());
    }

    if !bar.is_empty() {
        l1_parts.push(format!("{}{} {}%{}{}", clr, bar, pct, warn, RST));
    }
    if !cost_fmt.is_empty() {
        l1_parts.push(cost_fmt);
    }
    if !dur_fmt.is_empty() {
        l1_parts.push(dur_fmt);
    }
    if !git_display.is_empty() {
        let mut git_part = format!("{}{}{}", MAGENTA, git_display, RST);
        if !dirty.is_empty() {
            git_part = format!("{} {}{}{}", git_part, YELLOW, dirty, RST);
        }
        if !git_extra.is_empty() {
            git_part = format!("{} {}{}{}", git_part, CYAN, git_extra, RST);
        }
        l1_parts.push(git_part);
    }
    if !lines_fmt.is_empty() {
        l1_parts.push(lines_fmt);
    }

    let l1 = l1_parts.join(&sep);

    // ======== ASSEMBLE LINE 2 ========
    let l2 = if cfg.line2 {
        let mut l2_parts: Vec<String> = Vec::new();

        if cfg.show_tokens {
            let mut tok_parts: Vec<String> = Vec::new();
            if let Some(ref ms) = model_stats {
                if ms.opus_out > 0 || ms.opus_in > 0 {
                    tok_parts.push(format!(
                        "{}O{}:{}/{}",
                        MAGENTA,
                        RST,
                        format::fmt_k(ms.opus_in),
                        format::fmt_k(ms.opus_out)
                    ));
                }
                if ms.sonnet_out > 0 || ms.sonnet_in > 0 {
                    tok_parts.push(format!(
                        "{}S{}:{}/{}",
                        CYAN,
                        RST,
                        format::fmt_k(ms.sonnet_in),
                        format::fmt_k(ms.sonnet_out)
                    ));
                }
                if ms.haiku_out > 0 || ms.haiku_in > 0 {
                    tok_parts.push(format!(
                        "{}H{}:{}/{}",
                        GREEN,
                        RST,
                        format::fmt_k(ms.haiku_in),
                        format::fmt_k(ms.haiku_out)
                    ));
                }
            }
            if !tok_parts.is_empty() {
                l2_parts.push(tok_parts.join(" "));
            } else {
                l2_parts.push(format!(
                    "{}in:{}{} {}out:{}{}",
                    DIM, RST, in_fmt, DIM, RST, out_fmt
                ));
            }
        }

        if !speed_fmt.is_empty() {
            l2_parts.push(speed_fmt);
        }
        if !cum_proj.is_empty() {
            l2_parts.push(cum_proj);
        }
        if !cum_all.is_empty() {
            l2_parts.push(cum_all);
        }

        l2_parts.join(&sep)
    } else {
        String::new()
    };

    // --- NO_COLOR ---
    let (l1, l2) = if cfg.no_color {
        (strip_ansi(&l1), strip_ansi(&l2))
    } else {
        (l1, l2)
    };

    if !l2.is_empty() {
        format!("{}\n{}\n", l1, l2)
    } else {
        format!("{}\n\n", l1)
    }
}

/// Strip ANSI escape sequences manually (no regex dep).
fn strip_ansi(s: &str) -> String {
    let bytes = s.as_bytes();
    let mut result = Vec::with_capacity(bytes.len());
    let mut i = 0;
    while i < bytes.len() {
        if bytes[i] == 0x1b && i + 1 < bytes.len() && bytes[i + 1] == b'[' {
            // Skip until 'm'
            i += 2;
            while i < bytes.len() && bytes[i] != b'm' {
                i += 1;
            }
            if i < bytes.len() {
                i += 1; // skip 'm'
            }
        } else {
            result.push(bytes[i]);
            i += 1;
        }
    }
    String::from_utf8(result).unwrap_or_default()
}
